python/ray/data/dataset.py

```diff
@@ -318,88 +320,128 @@ def map_batches(
+        """Apply the given function to batches of data.
 
-        The format of the data batch provided to ``fn`` can be controlled via the
-        ``batch_format`` argument, and the output of the UDF can be any batch type.
+        Batches are represented as dataframes, ndarrays, or lists. The default batch
+        type is determined by your dataset's schema. To determine the default batch
+        type, call :meth:`~Dataset.default_batch_format`. Alternatively, set the batch
+        type with ``batch_format``.
 
-        This is a blocking operation.
+        To learn more about writing functions for :meth:`~Dataset.map_batches`, read
+        :ref:`writing user-defined functions <transform_datasets_writing_udfs>`.
+
+        .. note::
+            This is a blocking operation. The time complexity is :math:`O(n/p)`, where
+            :math:`n` is the dataset size and :math:`p` is the parallelism.
+
+        .. tip::
+            If you're using :ref:`Ray AIR <air>` for training or batch inference,
+            consider using :class:`~ray.data.preprocessors.BatchMapper`. It's more
+            performant and easier to use.
 
         Examples:
+
+            >>> import pandas as pd
             >>> import ray
-            >>> # Transform python objects.
-            >>> ds = ray.data.range(1000)
-            >>> # Transform batches in parallel.
-            >>> ds.map_batches(lambda batch: [v * 2 for v in batch])
-            Dataset(num_blocks=..., num_rows=1000, schema=<class 'int'>)
-            >>> # Define a callable class that persists state across
-            >>> # function invocations for efficiency.
-            >>> init_model = ... # doctest: +SKIP
+            >>> df = pd.DataFrame({
+            ...     "name": ["Luna", "Rory", "Scout"],
+            ...     "age": [4, 14, 9]
+            ... })
+            >>> ds = ray.data.from_pandas(df)
+            >>> ds
+            Dataset(num_blocks=1, num_rows=3, schema={name: object, age: int64})
+
+            Call :meth:`.default_batch_format` to determine the default batch
+            type.
+
+            >>> ds.default_batch_format()
+            <class 'pandas.core.frame.DataFrame'>
+
+            Once you know the batch type, define a function that transforms batches
+            of data. ``ds.map_batches`` applies the function in parallel.
+
+            >>> def map_fn(batch: pd.DataFrame) -> pd.DataFrame:
+            ...     batch["age_in_dog_years"] = 7 * batch["age"]
+            ...     return batch
+            >>> ds = ds.map_batches(map_fn)
+            >>> ds
+            Dataset(num_blocks=1, num_rows=3, schema={name: object, age: int64, age_in_dog_years: int64})
+
+            Your ``fn`` can return a different type than the input type. To learn more
+            about supported output types, read
+            :ref:`user-defined function output types <transform_datasets_batch_output_types>`.
+
+            >>> from typing import List
```

Comment:
Could you also add an example where the data is arrow blocks? Thanks!

Reply:
Do you mean an example where the underlying block type is Arrow, or the input batch format is Arrow?

Reply:
The input batch format is arrow

Reply:
I don't think you can get an Arrow input batch? AFAIK they always get converted to pandas.

```
def to_default(self) -> Block:
        if self.is_tensor_wrapper():
            default = self.to_numpy()
        else:
            # Always promote Arrow blocks to pandas for consistency, since
            # we lazily convert pandas->Arrow internally for efficiency.
            default = self.to_pandas()
        return default
```

Reply:
Oh, you mean like when you manually specify `batch_format=arrow`?

Reply:
Hi, Balaji, yes, you are right in that arrows get converted into pandas under the hood. However, as a user, I initially didn't know that and was looking for an example for the arrow input format. Maybe you can add a sentence somewhere to indicate that users with arrow input format can just use whatever panda uses since it's converted under the hood? 

The context is that I was using pymongoarrow [here](https://mongo-arrow.readthedocs.io/en/latest/quickstart.html) and was under the impression that the returned format is in arrow. Therefore, I wanted to use dataset to format it.

Tagging @jjyao  and @jianoaix  in case my description is unclear.

Reply:
> The context is that I was using pymongoarrow [here](https://mongo-arrow.readthedocs.io/en/latest/quickstart.html) and was under the impression that the returned format is in arrow. Therefore, I wanted to use dataset to format it.

Would it be correct to say that you were expecting the batch format to be an Arrow table because the underlying block type is Arrow?

My understanding is that the underlying block type is more of an implementation detail. Consumers of `read_*` APIs don't need to know the underlying block type; they just need to know just the batch format (related: https://github.com/ray-project/ray/pull/28434).

If the block type is an implementation detail, a sentence like "If the underlying block type is Arrow..." might be unhelpful or confusing to the majority of users.

@clarkzinzow what do you think?

Reply:
Hmm we already call out that the default batch format for tabular data (both Arrow and Pandas blocks) is the pandas DataFrame batch format in the docs for the `batch_format` parameter, we delve further into this in the "Transforming Datasets" feature guide, and we know have the `ds.default_batch_format()` API so users can query this at runtime, but maybe we need an explicit example or call out here for the Arrow --> Pandas case that makes this extra clear. But something akin to "Arrow blocks will yield pandas DataFrame batches by default" is also dipping into implementation details, as @bveeramani said, so not sure if we want to do that.

We have a few options here:
1. Have both a `ray.data.from_pandas()` example and a `ray.data.from_arrow()` example, showing that each yield a `pd.DataFrame` batch format by default.
2. Have a single example that's not tied to pandas or Arrow, e.g. `ray.data.read_parquet()` or `ray.data.read_csv()`, emphasizing that the abstract concept of "tabular" data is what leads to the default `pd.DataFrame` batch, which encompasses pandas DataFrames and Arrow Tables
3. Have a `ray.data.from_pandas()` example but have a sentence that a Dataset created from Arrow data will yield a `pd.DataFrame` batch by default (doesn't have to be a full example).
4. (2), and also have a sentence explicitly calling out that a Dataset created from Arrow data will yield a `pd.DataFrame` batch by default.

@shawnpanda Would all of these had made it clear that the default batch format for Arrow tabular data is still `pd.DataFrame`?

Reply:
Hi Balaji and Clark, thanks for the responses here

> Would all of these have made it clear that the default batch format for Arrow tabular data is still pd.DataFrame?

I have a preference for 1, 3, 4 so that the examples are more relatable to what I was working on and might be directly copy-able. The context is that I am very new to datasets and actually don't understand the nuances between different data formats. You guys can decide on a concise option and will defer to you guys. Don't want to make this great doc improvement PR feel like dragging on.